tests/src/core/testqgsexpression.cpp

```diff
@@ -1403,6 +1403,8 @@ class TestQgsExpression: public QObject
       QTest::newRow( "regexp_replace non greedy" ) << "regexp_replace('HeLLo','(?<=H).*?L', '-')" << false << QVariant( "H-Lo" );
       QTest::newRow( "regexp_replace cap group" ) << "regexp_replace('HeLLo','(eL)', 'x\\\\1x')" << false << QVariant( "HxeLxLo" );
       QTest::newRow( "regexp_replace invalid" ) << "regexp_replace('HeLLo','[[[', '-')" << true << QVariant();
+      QTest::newRow( "regexp_replace match, with null" ) << "regexp_replace('aaaa','a',NULL)" << false << QVariant( "" );
```

Comment:
```suggestion
      QTest::newRow( "regexp_replace match, with null" ) << "regexp_replace('aaaa','a',NULL)" << false << QVariant();
```

Reply:
Nope, it should return an empty string

Reply:
That seems counter intuitive to me. Wouldn't you need to use an empty string as the replacement character to get that? (How does postgres behave in this situation)

Reply:
@nyalldawson some tests on postgresql:

`SELECT regexp_replace('aaa', 'a', NULL) is NULL` => True
`SELECT regexp_replace('aaa', 'b', NULL) is NULL` => True
SELECT regexp_replace('abaa', 'a', NULL) is NULL => True
SELECT regexp_replace('abaa', 'b', NULL) is NULL => True
SELECT regexp_replace('abaa', 'c', NULL) is NULL => True

With our expression engine and Qt replace : `NULL` is considered as a `QVariant(Invalid)` and `str.replace` with `QString()` or `QString('')` will return the same string:

`QString("aaa").replace(QString("a"), QString())` => QString("")
`QString("aaabbb").replace(QString("b"), QString())` => QString("aaa")

So for you, if a NULL is sent as third parameter, we must return a `QVariant()`?

Reply:
I think if NULL is 3rd parameter AND the search string is found then yes, it should return null instead of empty string. If search string is NOT found then we should return original string.